pkg/compactor/block_upload.go

```diff
@@ -420,3 +491,180 @@ func (r bodyReader) ObjectSize() (int64, error) {
+		return blockIsComplete, nil, nil, nil
+	}
+
+	meta, err := c.loadUploadingMeta(ctx, userBkt, blockID)
+	if err != nil {
+		return blockStateUnknown, nil, nil, err
+	}
+	// If neither meta.json nor uploading-meta.json file don't exist, we say that block doesn't exist.
+	if meta == nil {
+		return blockUploadNotStarted, nil, nil, err
+	}
+
+	v, err := c.loadValidation(ctx, userBkt, blockID)
+	if err != nil {
+		return blockStateUnknown, meta, nil, err
+	}
+	if v == nil {
+		return blockUploadInProgress, meta, nil, err
+	}
+	if v.Error != "" {
+		return blockValidationFailed, meta, v, err
+	}
+	if time.Since(time.UnixMilli(v.LastUpdate)) < 5*time.Minute {
+		return blockValidationInProgress, meta, v, nil
+	}
+	return blockValidationStale, meta, v, nil
+}
+
+func (c *MultitenantCompactor) loadUploadingMeta(ctx context.Context, userBkt objstore.Bucket, blockID ulid.ULID) (*metadata.Meta, error) {
+	r, err := userBkt.Get(ctx, path.Join(blockID.String(), uploadingMetaFilename))
+	if err != nil {
+		if userBkt.IsObjNotFoundErr(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	defer func() { _ = r.Close() }()
+
+	v := &metadata.Meta{}
+	err = json.NewDecoder(r).Decode(v)
+	if err != nil {
+		return nil, err
+	}
+
+	return v, nil
+}
+
+func (c *MultitenantCompactor) loadValidation(ctx context.Context, userBkt objstore.Bucket, blockID ulid.ULID) (*validationFile, error) {
+	r, err := userBkt.Get(ctx, path.Join(blockID.String(), validationFilename))
+	if err != nil {
+		if userBkt.IsObjNotFoundErr(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	defer func() { _ = r.Close() }()
+
+	v := &validationFile{}
+	err = json.NewDecoder(r).Decode(v)
+	if err != nil {
+		return nil, err
+	}
+
+	return v, nil
+}
+
+func (c *MultitenantCompactor) uploadValidationWithError(ctx context.Context, logger log.Logger, blockID ulid.ULID,
+	userBkt objstore.Bucket, errorStr string) error {
+	val := validationFile{
+		LastUpdate: time.Now().UnixMilli(),
+		Error:      errorStr,
+	}
+	dst := path.Join(blockID.String(), validationFilename)
+	level.Debug(logger).Log("msg", fmt.Sprintf("uploading %s to bucket", validationFilename), "dst", dst)
+	buf := bytes.NewBuffer(nil)
+	if err := json.NewEncoder(buf).Encode(val); err != nil {
+		return errors.Wrap(err, "failed to encode block metadata")
+	}
+	if err := userBkt.Upload(ctx, dst, buf); err != nil {
+		return errors.Wrapf(err, "failed uploading %s to bucket", validationFilename)
+	}
+
+	return nil
+}
+
+func (c *MultitenantCompactor) uploadValidation(ctx context.Context, logger log.Logger, blockID ulid.ULID,
+	userBkt objstore.Bucket) error {
+	return c.uploadValidationWithError(ctx, logger, blockID, userBkt, "")
+}
+
+func (c *MultitenantCompactor) periodicValidationUpdater(ctx context.Context, logger log.Logger, blockID ulid.ULID,
+	userBkt objstore.Bucket, ch chan string, wg *sync.WaitGroup) {
+	defer wg.Done()
+	for {
+		select {
+		case <-ch:
+			return
+		case <-time.After(1 * time.Minute):
+			break
```

Comment:
We should not `break` here, but upload new validation. `break` means that `for` loop will stop, but we want `for` loop to continue in this case. `for` loop already stops if it receives notification via channel in `case <- ch`.